modules/lang-painless/src/test/java/org/elasticsearch/painless/GetByPathAugmentationTests.java

```diff
@@ -0,0 +1,183 @@
+        assertPathValue("['key0': ['a', 'b'], 'key1': ['c', 'd']]", "key1.0", "c");
+
+        List<String> k1List = new ArrayList<>();
+        k1List.add("c");
+        k1List.add("d");
+        assertPathValue("['key0': ['a', 'b'], 'key1': ['c', 'd']]", "key1", k1List);
+        assertPathValue("['k0': ['a','b','c','d'], 'k1': ['q']]", "k0.3", "d");
+
+        assertPathValue(mapMapList, "k0.k01.1", k001Obj);
+        assertPathValue(mapMapList, "k0.k01.1.k011", k001Value);
+
+
+        assertPathValue("[['key0': 'value0'], ['key1': 'value1']]", "1.key1", "value1");
+        assertPathValue("[['a','b'],['c','d'],[['e','f'],['g','h']]]", "2.1.1", "h");
+
+        assertPathValue(listMapListList, "2.m2.2", l2m2l1Obj);
+        assertPathValue(listMapListList, "2.m2.2.1", l2m2l1Index1);
+    }
+
+    public void testGetPathByValueDefaultValue() {
+        String mapList = "['key0': ['a', 'b'], 'key1': ['c', 'd']]";
+        assertPathDefaultValue(mapList, "key2", "x", "'x'");
+        assertPathDefaultValue(mapList, "key1.1", "d", "'x'");
+        assertPathDefaultValue(mapMapList, "k0.k01.1.k012", "foo", "'foo'");
+        assertPathDefaultValue(mapMapList, "k0.k01.1.k011", "b", "'foo'");
+        assertPathDefaultValue(mapMapList, "k0.k01.1", k001Obj, "'foo'");
+        assertPathDefaultValue(mapMapList, "k0.k01.9", k001Obj, k001MapStr);
+        assertPathDefaultValue(listMapListList, "2.m2.2", l2m2l1Obj, "'foo'");
+        assertPathDefaultValue(listMapListList, "2.m2.9", l2m2l1Obj, l2m2l1Str);
+    }
+
+    public void testGetPathByValueError() {
+        String path = "1.k0";
+        assertPathError("[['a','b'],['c','d']]", path, numberFormat("k0", path, 1));
+
+        path = "k0.k01.1.k012";
+        assertPathError(mapMapList, path, missingValue(path));
+
+        path = "k0.k01.k012";
+        assertPathError(mapMapList, path, numberFormat("k012", path, 2));
+
+        path = "2.m2.12";
+        assertPathError(listMapListList, path, missingValue(path));
+
+        path = "2.m2.a8";
+        assertPathError(listMapListList, path, numberFormat("a8", path, 2));
+
+        String container = "['a': ['b': 'c']]";
+        assertPathError(container, "a.b.c",
+            new IllegalArgumentException("Non-container [java.lang.String] at [c], index [2] in path [a.b.c]"));
+
+        assertPathError(container, "",
+            new IllegalArgumentException("Missing path"));
+
+        assertPathError(container, "a..b",
+            new IllegalArgumentException("Extra '.' in path [a..b] at index [1]"));
+
+        assertPathError(container, "a.b.",
+            new IllegalArgumentException("Trailing '.' in path [a.b.]"));
+    }
+
+    public void testGetByValueDefaultError() {
+        String path = "1.k0";
+        assertPathError("[['a','b'],['c','d']]", path, "'foo'", numberFormat("k0", path, 1));
+
+        path = "k0.k01.k012";
+        assertPathError(mapMapList, path, "'foo'", numberFormat("k012", path, 2));
```

Comment:
Each one of these assertions can be in its own test method. This makes reproductions much smaller and easier to deal with when debugging. For example, then when adding breakpoints, this makes it simpler to not have to deal with skipping over other assertions that is not the one being debugged.

Naming-wise, I would try to have descriptive but concise names. So we don't need the "GetByValue" in every test name, since that is what this entire test class is about. As an example, for this line one I might call it `testNonIntListIndexWithDefaultFails()`. The test name is also a form of documentation for future readers here, with the name giving some indication of what is trying to be tested.

The test method names would be something like `test

Reply:
That's much nicer, moved each to their own method.